add helper methods for detecting fs_insensitivty and comparing paths

hjanuschka opened the pull request:
trying  to fix: 
  - https://github.com/fastlane/fastlane/issues/11392
  - https://github.com/fastlane/fastlane/issues/10166

waiting for user confirmation.

hjanuschka committed:
add helper methods for detecting fs_insensitivty and comparing paths

hjanuschka committed:
fix

hjanuschka committed:
address feedback

hjanuschka commented:
user confirmed: https://github.com/fastlane/fastlane/issues/11392#issuecomment-355401903

KrauseFx commented on fastlane_core/lib/fastlane_core/helper.rb:
```diff
@@ -113,6 +113,25 @@ def self.is_test?
       self.test?
     end
 
+    def self.compare_paths(path1, path2)
```
Can you add some comments here on what this does, why it's necessary?

KrauseFx commented on fastlane_core/lib/fastlane_core/helper.rb:
```diff
@@ -113,6 +113,25 @@ def self.is_test?
+      compare_path1 = File.expand_path(path1)
+      compare_path2 = File.expand_path(path2)
+      if self.fs_is_insensitive?
+        # FS is insensitive, there fore the file `fastlane`  is also accessable via `FASTLANE``
```
I don't understand that comment, also double space after fastlane, also `therefore` I think?

KrauseFx commented:
Looks awesome, I understand the issue, but still needs some comments. You can even link the GitHub issues in the code base 👍

hjanuschka commented:
thx @KrauseFx feedback addressed

lacostej commented on fastlane_core/spec/helper_spec.rb:
```diff
@@ -58,6 +58,30 @@
+    end
+
+    describe "#compare_paths?" do
+      it "Compare Path on sensitive system" do
```
from a rspec style perspective, you could use `context` to avoid having to `expect` multiple times and make the intent clearer.

```
      context 'when the file system is insentive' do
        before(:each) do
          allow(FastlaneCore::Helper).to receive(:fs_is_insensitive?) { true }
        end
        it 'ignores case while comparing' do
             expect(FastlaneCore::Helper.compare_paths("FASTLANE", "fastlane")).to be true
        end
        it 'matches equal strings' do
             expect(FastlaneCore::Helper.compare_paths("fastlane", "fastlane")).to be true
        end
```

Not a big deal here, but might be useful in other "contextes" :)

lacostej commented on fastlane_core/lib/fastlane_core/helper.rb:
```diff
@@ -113,6 +113,36 @@ def self.is_test?
+
+    # Detect if filesystem is insenstive or not.
+    def self.fs_is_insensitive?
+      is_insensitive = false
```
ideally we could / should cache this value, especially as we create file+dir to resolve it.

hjanuschka committed:
feedback

hjanuschka commented:
@lacostej thx for your review, added a cache of the tmp file creation. and changed to the context in rspec, also added a bunch of new tests for the cache.

janpio commented:
Test failure on Ubuntu can be ignored as this was just a race condition thing with the documentation. Should be gone when rebasing on `master` though as we improved our testing setup.

Conflict should also go away.

@hjanuschka Can you rebase on master please? Then I think this is good to merge, @lacostej @KrauseFx ?

antoinelamy commented:
Bump, still experiencing the issue this PR aims to fix. @lacostej ? @KrauseFx ?

janpio commented:
This needs a rebase and conflict resolution before anyone can think about merging. You could possibly do so in a new branch and PR, I assume @hjanuschka just doesn't have the time right now.

antoinelamy commented:
@janpio done -> https://github.com/fastlane/fastlane/pull/13479
